bpfman-operator/apis/v1alpha1/bpf_application_types.go

```diff
@@ -0,0 +1,101 @@
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// BpfApplicationProgram defines the desired state of BpfApplication
+// +union
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'XDP' ?  has(self.xdpInfo) : !has(self.xdpInfo)",message="xdpInfo configuration is required when type is xdp, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'TC' ?  has(self.tcInfo) : !has(self.tcInfo)",message="tcInfo configuration is required when type is tc, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'TCX' ?  has(self.tcxInfo) : !has(self.tcxInfo)",message="tcxInfo configuration is required when type is tcx, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Uprobe' ?  has(self.uprobeInfo) : !has(self.uprobeInfo)",message="uprobeInfo configuration is required when type is uprobe, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'UretProbe' ?  has(self.uretprobeInfo) : !has(self.upretrobeInfo)",message="uretprobe configuration is required when type is uretprobe, and forbidden otherwise"
+type BpfApplicationProgram struct {
+	// name is the name of the function that is the entry point for the BPF
+	// program
+	Name string `json:"name"`
+
+	// type specifies the bpf program type
+	// +unionDiscriminator
+	// +required
+	// +kubebuilder:validation:Enum:="XDP";"TC";"TCX";"Uprobe";"UretProbe"
+	Type EBPFProgType `json:"type"`
+
+	// xdpInfo defines the desired state of the application's XdpPrograms.
+	// +unionMember
+	// +optional
+	XDPInfo *XdpProgramInfo `json:"xdpInfo,omitempty"`
+
+	// tcInfo defines the desired state of the application's TcPrograms.
+	// +unionMember
+	// +optional
+	TCInfo *TcProgramInfo `json:"tcInfo,omitempty"`
+
+	// tcxInfo defines the desired state of the application's TcxPrograms.
+	// +unionMember
+	// +optional
+	TCXInfo *TcxProgramInfo `json:"tcxInfo,omitempty"`
+
+	// uprobeInfo defines the desired state of the application's UprobePrograms.
+	// +unionMember
+	// +optional
+	UprobeInfo *UprobeProgramInfo `json:"uprobeInfo,omitempty"`
+
+	// uretprobeInfo defines the desired state of the application's UretprobePrograms.
+	// +unionMember
+	// +optional
+	UretprobeInfo *UprobeProgramInfo `json:"uretprobeInfo,omitempty"`
+}
+
+// BpfApplicationSpec defines the desired state of BpfApplication
+type BpfApplicationSpec struct {
+	BpfAppCommon `json:",inline"`
+
+	// programs is the list of bpf programs in the BpfApplication that should be
+	// loaded. The application can selectively choose which program(s) to run
+	// from this list based on the optional attach points provided.
+	// +kubebuilder:validation:MinItems:=1
+	Programs []BpfApplicationProgram `json:"programs,omitempty"`
```

Comment:
Is this meant to be an optional field? What happens if I create a `BpfApplication` and specify no programs?

Reply:
Generally for lists we recommend having an upper bound to the number of entries. Is there a reasonable upper bound that can be specified here?

Additionally, what other constraints exist here? Can I have multiple program entries with the same `.name` value?

Reply:
The purpose of a BpfApplication is to ask bpfman to load and possibly attach eBpf programs.  I don't see a use case for a BpfApplication CRD with no programs identified.  Regarding the bound, I don't know what an acceptable bound would be.  The same goes for links.

Reply:
In that case, this should probably be a required field. For an upper bound, how many programs would you generally expect a user to specify? What number would surprise you as being a really high number?

Reply:
Agreed, it should be required.  We'll fix that.

Reply:
I didn't want to arbitrarily limit it.  However, we'll try to come up with some upper bounds on what's reasonable.

Reply:
You can always increase it in the future if needed. You can never decrease it though.

If you _truly_ can't enforce a reasonable upper bound then that is fine, but usually there is a reasonable one that can be set